### PR TITLE
Set crow_indices outputs as non-differentiable.

### LINKED ISSUE #107083: Getting indices of a sparse compressed tensor results tensors with requires_grad being True
## Issue description

As in the title.

It is a bug because "only Tensors of floating point dtype can require gradients".

## Code example

```python
>>> a = torch.tensor([[1, 2], [3, 4]], dtype=torch.float64, requires_grad=True).to_sparse_csr()
>>> a
tensor(crow_indices=tensor([0, 2, 4]),
       col_indices=tensor([0, 1, 0, 1]),
       values=tensor([1., 2., 3., 4.]), size=(2, 2), nnz=4,
       dtype=torch.float64, layout=torch.sparse_csr,
       grad_fn=<ToSparseCsrBackward0>)
>>> a.crow_indices()
tensor([0, 2, 4], requires_grad=True)
```

Expected `a.crow_indices().requires_grad == False`.

## System Info

- PyTorch version: main

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer

### PR DESCRIPTION
**Pull request by pearu**
Fixes https://github.com/pytorch/pytorch/issues/107083

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107150
* #107384
* __->__ #107447

cc @ezyang @albanD @zou3519 @gqchen @nikitaved @soulitzer @Lezcano @Varal7

### CODE REVIEW

File: aten/src/ATen/core/TensorBase.h [hunk 1]
```diff
@@ -741,7 +741,7 @@ class TORCH_API TensorBase {
     return *this;
   }
   bool requires_grad() const {
-    return impl_->requires_grad();
+    return impl_->requires_grad() && (is_floating_point() || is_complex());
```
Comment (pearu) [inline aten/src/ATen/core/TensorBase.h:744]: In the case of CSR tensor `crow_indices()` return value (that is a tensor with integer dtype), `impl_->requires_grad()` may return `true` when `impl_->autograd_meta_->requires_grad()` returns true.

So, an alternative solution is to make sure that `autograd_meta_` is not set for `crow_indices()` return value. Please advise if this would be a preferred solution and how to achieve this.

^ @albanD

Comment (pearu) [reply aten/src/ATen/core/TensorBase.h]: Nevermind, adding
```
- name: crow_indices(Tensor(a) self) -> Tensor(a)
  output_differentiability: [False]
```
to derivatives.yaml resolved the problem.

**Review by cpuhrsch — APPROVED** [review]
